lib/rs/src/errors.rs

```diff
@@ -0,0 +1,678 @@
+///       ApplicationErrorKind::Unknown,
+///       format!("This is an error")
+///     )
+///   )
+/// );
+/// ```
+///
+/// Returning a user-defined (using the Thrift IDL) exception struct.
+///
+/// ```text
+/// // Thrift IDL exception definition.
+/// exception Xception {
+///   1: i32 errorCode,
+///   2: string message
+/// }
+/// ```
+///
+/// ```
+/// use std::convert::From;
+/// use std::error::Error;
+/// use std::fmt;
+/// use std::fmt::{Display, Formatter};
+///
+/// // auto-generated by the Thrift compiler
+/// #[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+/// pub struct Xception {
+///   pub error_code: Option<i32>,
+///   pub message: Option<String>,
+/// }
+///
+/// // auto-generated by the Thrift compiler
+/// impl Error for Xception {
+///   fn description(&self) -> &str {
+///     "remote service threw Xception"
+///   }
+/// }
+///
+/// // auto-generated by the Thrift compiler
+/// impl From<Xception> for thrift::Error {
+///   fn from(e: Xception) -> Self {
+///     thrift::Error::User(Box::new(e))
+///   }
+/// }
+///
+/// // auto-generated by the Thrift compiler
+/// impl Display for Xception {
+///   fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+///     self.description().fmt(f)
+///   }
+/// }
+///
+/// // in user code...
+/// let err: thrift::Result<()> = Err(
+///   thrift::Error::from(Xception { error_code: Some(1), message: None })
+/// );
+/// ```
+pub enum Error {
+    /// Errors encountered while writing byes to the I/O channel.
+    ///
+    /// These include "connection closed", "bind errors", etc.
+    Transport(TransportError),
+    /// Errors encountered when processing a Thrift message within the rift
+    /// library.
+    ///
+    /// These include "exceeding size limits", "unsupported protocol versions",
+    /// etc.
+    Protocol(ProtocolError),
+    /// Errors encountered when Thrift-compiler-generated 'application' code
+    /// encounters conditions that violate the spec.
+    ///
+    /// These include "out-of-order messages", "missing required-field values",
+    /// etc. This variant also functions as a catch-all: any error returned by
+    /// handler functions is automatically encoded as an `ApplicationError` and
+    /// returned to the caller.
+    Application(ApplicationError),
+    /// Carries exception structs defined within the Thrift IDL.
+    User(Box<error::Error + Sync + Send>),
+}
+
+impl Error {
+    /// Read the wire representation of an application exception thrown by the
+    /// remote Thrift endpoint into its corresponding rift `ApplicationError`
+    /// representation.
+    ///
+    /// Application code should never call this method directly.
+    pub fn read_application_error_from_in_protocol(i: &mut TInputProtocol)
+                                                   -> ::Result<ApplicationError> {
+        let mut message = "general remote error".to_owned();
+        let mut kind = ApplicationErrorKind::Unknown;
+
+        try!(i.read_struct_begin());
```

Comment:
The [doc](https://doc.rust-lang.org/std/macro.try.html) says

```
Prefer using ? syntax to try!. ? is built in to the language and is more succinct than try!. It is the standard method for error propagation.
```

So `?` is more Rust-idiomatic and future-proof way to check the result.